Mathlib/RingTheory/LinearDisjoint.lean

```diff
@@ -0,0 +1,603 @@
+
+/-- If a basis of `A` is also `B`-linearly independent, if elements in `A` and `B` commute,
+then `A` and `B` are linearly disjoint. -/
+theorem of_basis_left_of_commute {ι : Type*} (a : Basis ι R A)
+    (H : LinearIndependent B (A.val ∘ a)) (hc : ∀ (a : A) (b : B), Commute a.1 b.1) :
+    A.LinearDisjoint B :=
+  (of_basis_right B A a H).symm_of_commute fun _ _ ↦ (hc _ _).symm
+
+variable {A B} in
+/-- If `A` and `B` are linearly disjoint, if `A` is flat, then for any family of
+`R`-linearly independent elements `{ a_i }` of `A`, and any family of
+`R`-linearly independent elements `{ b_j }` of `B`, the family `{ a_i * b_j }` in `S` is
+also `R`-linearly independent. -/
+theorem linearIndependent_mul_of_flat_left (H : A.LinearDisjoint B) [Module.Flat R A]
+    {κ ι : Type*} {a : κ → A} {b : ι → B} (ha : LinearIndependent R a)
+    (hb : LinearIndependent R b) : LinearIndependent R fun (i : κ × ι) ↦ (a i.1).1 * (b i.2).1 := by
+  haveI : Module.Flat R (toSubmodule A) := ‹_›
+  exact H.1.linearIndependent_mul_of_flat_left ha hb
+
+variable {A B} in
+/-- If `A` and `B` are linearly disjoint, if `B` is flat, then for any family of
+`R`-linearly independent elements `{ a_i }` of `A`, and any family of
+`R`-linearly independent elements `{ b_j }` of `B`, the family `{ a_i * b_j }` in `S` is
+also `R`-linearly independent. -/
+theorem linearIndependent_mul_of_flat_right (H : A.LinearDisjoint B) [Module.Flat R B]
+    {κ ι : Type*} {a : κ → A} {b : ι → B} (ha : LinearIndependent R a)
+    (hb : LinearIndependent R b) : LinearIndependent R fun (i : κ × ι) ↦ (a i.1).1 * (b i.2).1 := by
+  haveI : Module.Flat R (toSubmodule B) := ‹_›
+  exact H.1.linearIndependent_mul_of_flat_right ha hb
+
+variable {A B} in
+/-- If `A` and `B` are linearly disjoint, if one of `A` and `B` is flat, then for any family of
+`R`-linearly independent elements `{ a_i }` of `A`, and any family of
+`R`-linearly independent elements `{ b_j }` of `B`, the family `{ a_i * b_j }` in `S` is
+also `R`-linearly independent. -/
+theorem linearIndependent_mul_of_flat (H : A.LinearDisjoint B)
+    (hf : Module.Flat R A ∨ Module.Flat R B)
+    {κ ι : Type*} {a : κ → A} {b : ι → B} (ha : LinearIndependent R a)
+    (hb : LinearIndependent R b) : LinearIndependent R fun (i : κ × ι) ↦ (a i.1).1 * (b i.2).1 :=
+  H.1.linearIndependent_mul_of_flat hf ha hb
+
+/-- If `{ a_i }` is an `R`-basis of `A`, if `{ b_j }` is an `R`-basis of `B`,
+such that the family `{ a_i * b_j }` in `S` is `R`-linearly independent,
+then `A` and `B` are linearly disjoint. -/
+theorem of_basis_mul {κ ι : Type*} (a : Basis κ R A) (b : Basis ι R B)
+    (H : LinearIndependent R fun (i : κ × ι) ↦ (a i.1).1 * (b i.2).1) : A.LinearDisjoint B :=
+  ⟨.of_basis_mul _ _ a b H⟩
+
+variable {A B}
+
+variable (H : A.LinearDisjoint B)
+
+theorem of_le_left_of_flat {A' : Subalgebra R S}
+    (h : A' ≤ A) [Module.Flat R B] : A'.LinearDisjoint B := by
+  haveI : Module.Flat R (toSubmodule B) := ‹_›
```

Comment:
Can you make this into an instance? (`instance [Module.Flat R A] : Module.Flat R A.toSubmodule`)

Reply:
Do you mean make it global for mathlib (which I think will have a performance impact), or only locally in this file?

Reply:
I don't think there will be a performance impact.

Reply:
PR created as #15360.

Reply:
Done.